pkgs/applications/virtualization/singularity/template.nix

```diff
@@ -0,0 +1,187 @@
+{ lib
+, util-linux
+, gpgme
+, openssl
+, libuuid
+, bash
+, coreutils
+, which
+, makeWrapper
+, cryptsetup
+, squashfsTools
+, nvidia-docker # For nvidia-container-cli
+, glibc # For ldconfig needed by nvidia-container-cli
+, go
+, buildGoModule
+, pname
+, version
+, src
+, projectName # "apptainer" or "singularity"
```

Comment:
> * name their "template" as `generic.nix`

That sounds quite common. I'll do the renaming after addressing the above suggestions.


> * their template has the form `{ version, variant, etc }: { lib, stdenv, actualOverridableBuildInputs, etc }: ...`, where the first part of the arguments apparently is not overriddable with `.override`

After doing a `Ctrl+r` searching with the VSCode, it seems that this isn't a unified practice. Besides, `callPackage ({version}: { lib, stdenv }: stdenv.mkDerivation { ...; }) { version = "0.1.0"; } { }` will cause the resulting package to loose the `.override` functionality. (calling the first layer and then do `callPackage` may solve the problem, but I haven't try that before.) After putting the version stuff into the first attrset, we still need to do the overriding due to the `buildGoModule` issue #86349.

Reply:
> Here's another question I have, simply because I'm unaware: it seems like a convention, that expressions shipping multiple versions
> 
>     * name their "template" as `generic.nix`,
> 
>     * their template has the form `{ version, variant, etc }: { lib, stdenv, actualOverridableBuildInputs, etc }: ...`, where the first part of the arguments apparently is not overriddable with `.override`
> 
> 
> I have no clue why people do that, nor if it's an actual convention rather than cargo cult - but I'd probably find answers prior to introducing the new pattern. What do you think?

Solved.